### ISSUE TITLE
The HeaderEditor doesn't work on the Basilisk browser

### ISSUE THREAD

**Issue by Telmesomething** [FirefoxBar/HeaderEditor#44]
[Basilisk](https://www.basilisk-browser.org/releasenotes.shtml) is a new Firefox fork released a couple of weeks ago which supports XUL (legacy) extensions. In response to an article on [ghacks.net](https://www.ghacks.net/2017/12/09/a-solution-to-etag-tracking-in-firefox/) I installed your extension this evening, but unfortunately it isn't possible to save any rules. The toolbar button doesn't work and using the Add-ons Manager/Options menu to open the menu allows a rule to be created, but nothing happens when clicking the Save button. 

I tried using an older version but that didn't help either. 

If possible, please fix the problem.

**Comment by sylingd** [comment]
Please open console (Press F12) in management page, and take a screenshot to me

**Comment by Telmesomething** [comment]
Here you are: 

![console_header_editor](https://user-images.githubusercontent.com/33827636/33803026-7ca46612-dd86-11e7-8810-54a712f6cbbf.PNG)

**Comment by sylingd** [comment]
It seems that Basilisk has tighter restrictions on security (Or it has a BUG), But I did not find any documention about it.

**Comment by Thorin-Oakenpants** [comment]
Just FYI to those who read this

- Are you in Private Browsing mode? (PB mode blocks use of indexedDB where the rules are stored)
   * If so, set/edit your rules etc from a normal window, and I *think* your PB Mode will still "read" from the web extension's IDB entries (but don't quote me on that)
- Do you have any of the following settings
   * Options>Privacy&Security>History>Firefox will **never remember history**
   * Options>Privacy&Security>History>custom settings> never accept any cookies

Cookies control access to localstorage (not the web extension's Storage API) and IDB. This is a bug with web extensions (see [1406675](https://bugzilla.mozilla.org/show_bug.cgi?id=1406675) which should be fixed in FF59). As for PB mode, IDB is blocked by design.

If the problem in your cookie settings, then you can still set an extension cookie permission exception - see [here](https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1.1-Setting-Extension-Permission-Exceptions) - warning, comes with some helpful pictures :)

@Telmesomething : due to the error description, I don't think this is the case for you, but then again, who knows. Can you confirm you're not afflicted by the above wall of text I just typed out :)

**Comment by Telmesomething** [comment]
> @Telmesomething : due to the error description, I don't think this is the case for you, but then again, who knows. Can you confirm you're not afflicted by the above wall of text I just typed out :)
I wasn't using any of the settings you mentioned when trying to create a rule. 

I checked in C:\Users\{Username}\AppData\Roaming\Moonchild Productions\Basilisk\Profiles\xxxxx.default\storage but couldn't find any IDB for the Header Editor in there. Should there be one?

I have Self-Destructing Cookies installed, but suspended the operation while testing your extension, but it didn't make any difference. 

I've managed to find an alternative extension called [Modifying Headers](https://addons.mozilla.org/en-US/firefox/addon/modify-headers/) though and used that to create a filter called Remove Etag which seems work with Basilisk. I tested it on [Cookieless cookies](http://lucb1e.com/rp/cookielesscookies/) so I figure there must be another reason why yours doesn't work although I don't have a clue what that might be I'm afraid. Here's a screenshot of the rule created on MH in case it helps.

![modify_headers](https://user-images.githubusercontent.com/33827636/33832179-5873f3c0-de7b-11e7-80f6-05724c78d4fc.PNG)

**Comment by Telmesomething** [comment]
> I just downloaded Basilisk and it works well

I just this minute created a new profile with only Header Editor installed and like you say, it works. So I guess I owe you an apology. 

The problem is likely due to an about:config change I've made somewhere. So I'll have to dump my previous profile, install all the extensions that were present in previous one and then without making any about:config changes, test whether Header Editor is still working properly or not. That will at least confirm that it isn't a conflict with another extension. 

Anyway, my sincere apologies for wasting your time like that. I'll make a mental note to test extensions in a new profile before posting about a problem which might not actually exist in future.

**Comment by sylingd** [comment]
I guess you have change some configs about privacy, you can check them. For example, open private mode will make WebExtension can not access IndexedDB.

**Comment by Telmesomething** [comment]
> I guess you have change some configs about privacy, you can check them. For example, open private mode will make WebExtension can not access IndexedDB.

Ok, thanks for the tip.

But you emailed me a link to a tutorial on how to add a cookie permission exception using about:debugging. I wanted to have a go at that, but immediately ran into a problem. On the about:debugging#addons page, there's just a long list of the extensions I have installed each of which has a Debug and a Reload button to the right of it. Here's a pix of yours

![about-debugging](https://user-images.githubusercontent.com/33827636/33843720-6a8834e0-de9e-11e7-9ee9-dad058a4ddaa.PNG)

Nothing is clickable until the top box in checkmarked and after which only the Debug button becomes available. I tried clicking that but was was presented with the warning you see in the next pix.

![incoming-debugging](https://user-images.githubusercontent.com/33827636/33843831-bff00a02-de9e-11e7-8c87-a44546977b2c.PNG)

Since both endpoints were loopback addresses I went ahead and allowed the connection, but that just loads the console and there doesn't appear to be any way of loading the manifest URL mentioned in the tutorial in any of the options. 

Any ideas on that one?

**Comment by Thorin-Oakenpants** [comment]
> Forgot to mention that I copy/pasted the Header Editor moz.xxxxxxxx extension in the storage\default folder in the new profile into the old one at the same location, but the menu to add a rule still doesn't work.

the `xxxx` part is a UUID unique to each extension installation (although you can keep it per profile - eg un-install the extension, re-install it etc to troubleshoot, but it relies on a pref), so can't do that - but we're talking Basilik, not FF here, so do not quote me on that

**Comment by sylingd** [comment]
If you can not see UUID in debugging page, you can open management page and look at the address bar
![qq 20171212144836](https://user-images.githubusercontent.com/5326684/33871010-96b8bc58-df4b-11e7-9a61-9f2f150fedb8.png)

**Comment by Telmesomething** [comment]
> the xxxx part is a UUID unique to each extension installation (although you can keep it per profile - eg un-install the extension, re-install it etc to troubleshoot, but it relies on a pref), so can't do that - but we're talking Basilik, not FF here, so do not quote me on that

OK, understood.

**Comment by Telmesomething** [comment]
> If you can not see UUID in debugging page, you can open management page and look at the address bar

I can see a tab called Firefox Developer in your screenshot. But which of the options in the screenshot do I have to enable in order to reach that page?

![developer_tools](https://user-images.githubusercontent.com/33827636/33874724-085ada98-df20-11e7-9902-d1536dc07885.PNG)

**Comment by sylingd** [comment]
It is about:debugging, I think you have opened this page but it is different from the offical version of Firefox

**Comment by Telmesomething** [comment]
> It is about:debugging, I think you have opened this page but it is different from the offical version of Firefox

Yes, I was afraid you would say that. I was rather hoping there was another option somewhere in the Web Developer tools. 

But I have the UUID already, but don't quite understand how I'm supposed to use it to create a cookie exception. I should add here that although it appears in the location bar, the IDB is nowhere to be found in the profile storage\default directory.

**Comment by Thorin-Oakenpants** [comment]
> But I have the UUID already, but don't quite understand how I'm supposed to use it to create a cookie exception

See - https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1.1-Setting-Extension-Permission-Exceptions#small_orange_diamond-note .. see the picture, you can manually add it - eg:
`moz-extension://8a1a8222-6b59-4639-b488-48ef03fa3b57`

**Comment by Telmesomething** [comment]
So I managed to add the moz-extension as an exception, but still can't enable any rules. Both `dom.indexedDB.enabled` and `dom.storage.enable` are set to true though. 

I have the following privacy settings implemented in about: config

`privacy.firstparty.isolate = true`
`network.http.referer.spoofSource = true `
`browser.safebrowsing.malware.enabled = false`
`beacon.enabled = false`
`media.navigator.enabled = false`
`extensions.getAddons.cache.enabled = false`
`dom.event.clipboardevents.enabled = false`
`geo.enabled = false`
`webgl.disabled = true`
`media.peerconnection.enabled = false`
`network.cookie.cookieBehavior = 1`

I have [Smart Referer](https://addons.mozilla.org/en-US/firefox/addon/smart-referer/)  installed to handle referer options. 

And [Self-Destructing Cookies](https://addons.mozilla.org/en-US/firefox/addon/self-destructing-cookies/) is set to allow session cookies for sites I commonly use like Github, banking etc.. It's also configured to clear the browser cache every minute. 

Thanks for the other link though. I don't use Private Browsing that much actually and not at all for normal everyday use such as visiting news media sites which makes up the bulk of my time on the Web.

**Comment by Telmesomething** [comment]
Aha! So it was that last one in the list (above) which was set to 2 (block all network cookies). I really don't recall setting it to that value though and assume the setting was implemented by one of my other extensions. 

So now Header Editor works!

**Comment by Thorin-Oakenpants** [comment]
yup .. cookies control access to localstorage and IDB .. just as well, as I don;t have time to fluff around with basilisk - I DL'ed it, unpacked it and stopped there :) More time for beersies

**Comment by Telmesomething** [comment]
Just a quickie. The `network.cookie.cookieBehavior = 2` setting came from the prefs page. Changing it to "1" serves to re-enable Accept cookies from sites again which is likely to screw up the Self-Destructing Cookies extension settings. So I've reset it to "2" once more. but at least I know where the problem originated from.

**Comment by Telmesomething** [comment]
> yup .. cookies control access to localstorage and IDB .. just as well, as I don;t have time to fluff around with basilisk - I DL'ed it, unpacked it and stopped there :) More time for beersies

I just confirmed that resetting `network.cookie.cookieBehavior = 2` has removed the rule I created which is a pity since I was rather hoping that would remain in place. I don't really like the idea of allowing even first party cookies to sites due to tracking issues. Fortunately though, the other extension I mentioned earlier serves to delete Etags even with cookies disabled so I've re-enabled that. 

Thanks once again for all the help you provided though. It's been an education.
